doc/guides/pm/power_domain.rst

```diff
@@ -0,0 +1,184 @@
+
+      action -> devA [label="pm_device_get()"]
+      devA:se -> domain:n [label="pm_device_get()"]
+
+      domain -> devB [label="action_cb(PM_DEVICE_ACTION_TURN_ON)"]
+      domain:sw -> devA:sw [label="action_cb(PM_DEVICE_ACTION_TURN_ON)"]
+   }
+
+Internal Power Domains
+----------------------
+
+Most of the devices in an SoC have independent power control that can
+be turned on or off to reduce power consumption. But there is a
+significant amount of static current leakage that can't be controlled
+only using device power management. To solve this problem, SoCs
+normally are divided into several regions grouping devices that
+are generally used together, so that an unused region can be
+completely powered off to eliminate this leakage. These regions are
+called "power domains", can be present in a hierarchy and can be
+nested.
+
+External Power Domains
+----------------------
+
+Devices external to a SoC can be powered from sources other than the main power
+source of the SoC. These external sources are typically a switch, a regulator,
+or a dedicated power IC. Multiple devices can be powered from the same source,
+and this grouping of devices is typically called a "power domain".
+
+Placing devices on power domains can be done for a variety of reasons,
+including to enable devices with high power consumption in low power mode to be
+completely turned off when not in use.
+
+Implementation guidelines
+*************************
+
+In a first place, a device that acts as a power domain needs to
+declare compatible with ``power-domain``. Taking
+:ref:`pm-domain-work-flow` as example, the following code defines a
+domain called ``gpio_domain``.
+
+.. code-block:: devicetree
+
+	gpio_domain: gpio_domain@4 {
+		compatible = "power-domain";
+		...
+	};
+
+A power domain needs to implement the PM action callback used by the
+PM subsystem to turn devices on and off.
+
+.. code-block:: c
+
+    static int mydomain_pm_action(const struct device *dev,
+                               enum pm_device_action *action)
+    {
+        switch (action) {
+        case PM_DEVICE_ACTION_RESUME:
+            /* resume the domain */
+            ...
+            /* notify children domain is now powered */
+            pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+            break;
+        case PM_DEVICE_ACTION_SUSPEND:
+            /* notify children domain is going down */
+            pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+            /* suspend the domain */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_ON:
+            /* turn on the domain (e.g. setup control pins to disabled) */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_OFF:
+            /* turn off the domain (e.g. reset control pins to default state) */
+            ...
+            break;
+        default:
+            return -ENOTSUP;
+        }
+
+        return 0;
+    }
```

Comment:
could the domain workflow be clarified now? When/how is the domain turned off/on, and what does this imply? what are the differences between resume/suspend/turn_on/turn_off?

Reply:
AFAIU these operations are only triggered when the domain is nested in another domain that was suspended / resumed.